### PR TITLE
[AIP-131] Lint the google.api.http annotation

### PR DESCRIPTION
**Pull request by lukesneeringer**
This adds linting of the `google.api.http` annotation on Get requests.
It also adds a friendly "getHTTPRules" function to range over the http declarations.

### CODE REVIEW

File: rules/0131.go [hunk 1]
```diff
@@ -71,6 +73,84 @@ func checkGetRequestMessageName() lint.Rule {
...
+					return
+				}
+
+				// Iterate over the HTTP rules.
+				for _, httpRule := range getHTTPRules(m) {
+					// Rule check: Ensure that the GET HTTP verb is used.
+					getURI := httpRule.GetGet()
+					if getURI == "" {
+						problems = append(problems, lint.Problem{
+							Message:    "Get RPCs should use the GET HTTP verb.",
```
Comment (alin04) [inline rules/0131.go]: I haven't been too thorough. Are you using "should" vs "must" as appropriate in these error messages?

File: rules/utils.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+package rules
+
+import (
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/genproto/googleapis/api/annotations"
+	p "google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+// GetHTTPRules returns a array of HTTP rules for a given method descriptor.
```
Comment (alin04) [inline rules/utils.go]: getHTTPRules (i.e. lowercaseCamelCase) to match method name

File: rules/utils.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
+// Copyright 2019 Google LLC
```
Comment (alin04) [inline rules/utils.go]: utils.go is pretty generic, and could become a huge dumping ground. Maybe name this file http_annotations.go to keep all utilities around annotations co-located?

File: rules/0131.go (hunk 1 above)
Comment (lukesneeringer) [reply rules/0131.go]: Probably not. Good call.

File: rules/utils.go (hunk 3 above)
Comment (lukesneeringer) [reply rules/utils.go]: I have almost the opposite concern -- given that this is a `rules` package, the number of things that are not rules should be as limited as possible.

File: rules/utils.go (hunk 2 above)
Comment (lukesneeringer) [reply rules/utils.go]:
```suggestion
// getHTTPRules returns a array of HTTP rules for a given method descriptor.
```

File: rules/0131.go (hunk 1 above)
Comment (lukesneeringer) [reply rules/0131.go]: This one definitely should be must.

```suggestion
							Message:    "Get RPCs must use the GET HTTP verb.",
```

File: rules/utils.go (hunk 3 above)
Comment (mingzhi) [reply rules/utils.go]: Maybe you want to create an internal package for util: `internal/util` under `rules`.

Comment (lukesneeringer) [reply rules/utils.go]: That makes me unable to read the `testdata`.

File: rules/utils_test.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		fdp := testutil.MustCreateFileDescriptorProto(
+			t,
+			testutil.FileDescriptorSpec{
+				AdditionalProtoPaths: []string{testdatadir("api-common-protos")},
+				Filename:             "test.proto",
+				Template:             tmpl,
+				Data:                 test,
+			},
+		)
+		reg, err := lint.MakeRegistryFromAllFiles([]*descriptorpb.FileDescriptorProto{fdp})
```
Comment (robcapo) [inline rules/utils_test.go]: We shouldn't expose this method just for the sake of this test. This functionality doesn't really even belong in the linter – I wrote it as a temporary helper until the proto team adds this feature to protodesc. Try using [protoregistry.NewFiles](https://github.com/golang/protobuf/blob/api-v2/reflect/protoregistry/registry.go#L76) directly.

Maybe you can make a new `testutil` method that takes a directory path and outputs a `[]*descriptorpb.FileDescriptorProto`. Or maybe there's some way to get the MethodDescriptor from the descriptorpb version of it.

Comment (lukesneeringer) [reply rules/utils_test.go]: `protoregistry.NewFiles` takes a `protoreflect.FileDescriptor`, which is not the thing I have.

Comment (robcapo) [reply rules/utils_test.go]: Yeah, the coordination between the two packages can be a bit annoying, but I still feel that this function should not be exposed from our package. We intend on removing it. I would almost prefer you copy paste it as a private function into this package if you can't find a cleaner way to test your helper.

**Comment by lukesneeringer** [comment]
Replaced by #188.

**Pull request closed without merging**